biz.aQute.bndlib/src/aQute/bnd/build/WorkspaceLock.java

```diff
@@ -103,6 +104,30 @@ private TimeoutException timeout(Lock lock) {
 		return e;
 	}
 
+	<T, U> T writeReadLocked(final long timeoutInMs, final Callable<U> underWrite,
+		final FunctionWithException<U, T> underRead, final BooleanSupplier canceled) throws Exception {
+		Callable<U> writeLocked = () -> {
+			U writeResult = underWrite.call();
+			// It is safe to let an exception propagate from here.
+			// Downgrade lock by taking read lock before returning to release
+			// write lock.
+			trace("Downgrading", writeLock());
+			readLock().lock();
+			return writeResult;
+		};
+		U writeResult = locked(writeLock(), timeoutInMs, writeLocked, canceled);
+		// If the call to locked completed normally, we hold the read
+		// lock obtained in the writeLocked callable.
+		try {
+			Callable<T> readLocked = () -> underRead.apply(writeResult);
+			return locked(readLock(), timeoutInMs, readLocked, canceled);
```

Comment:
? We know we have a read lock, why would we try to lock it again? Unless canceled, we can directly call the underRead lambda? We do not need 2 read locks? 

The timeout is now questionable, would it need to be adjusted? However it should always succeed since we have a read lock.

Reply:
This is using the existing locked method which has tracing, interrupt handling etc. already in place. Better to use the one method rather than another code path. Taking 2 counts on the read lock is no big deal. The timeout is meaning less as we already hold the read lock, so we well not be delaying in getting it again.